Compiler/test/Spec.hs

```diff
@@ -1,37 +1,79 @@
+import Control.Monad.State
+
 import Test.Hspec
 
 import Text.Parsec
 
 import AST
+import Env
+import Preprocessor
 import Parser
 
 main :: IO ()
 main = hspec $ do
+    preprocessorTests
     parserTests
 
+preprocessorTests = do
```

Comment:
The preprocessor tests are difficult to read as is, since they are being built directly using data constructors. Consider calling the parser to build pre-conditions instead, or maybe adding comments with what the actual program is so it is clear what is going on.